### ISSUE TITLE
GPU inference on pretrained NER model goes OOM after processing several documents

### ISSUE THREAD

**Issue by HonzaCech** [explosion/spaCy#9432]
## How to reproduce the behaviour
I am using `en_core_web_trf` to extract named entitites from documents. I send the data to the model as string - single string per document. Documents are usually 2-10 pages long english texts.  After processing several documents - 5,6, mostly under 10 - we get the OOM error from CUDA. I have posted a [question](https://stackoverflow.com/questions/69301276/spacy-inference-goes-oom-when-processing-several-documents) on SO and people there helped me identify the problem is that the transformer inside keeps the gradient and accumulates it over multiple inferences. They've even given me a working solution for this, but I believe it might be a spacy bug, as I don't suppose that should happen.
```class SpacyExtractor():
    def __init__(self):
        spacy.require_gpu()
        self.model = spacy.load('en_core_web_trf',
                                 disable=["tagger", "parser", "attribute_ruler", "lemmatizer"])


    def get_named_entities(self, text: str):
        doc = self.model(text)
        entities = []
        for ent in doc.ents:
            entities.append((ent.text, ent.label_))
        return entities
```
With this code, I repeatedly call `get_named_entities` with strings I receive (via REST api). I tried checking allocated memory by nvidia-smi after each document and the allocated memory reallly grew up after each call, before finally getting OOM error.

It can be solved by doing
```
with torch.no_grad():
    doc = self.model(text)
```

## Your Environment
- **spaCy version:** 3.1.3
- **Platform:** Ubuntu 18.04
- **Python version:** 3.8.0
- **Pipelines:** en_core_web_md (3.1.0), en_core_web_trf (3.1.0)
- **Cuda:** 9.1
- **GPU:** RTX 2080 Ti 11GB

**Comment by martin-kirilov** [comment]
I can confirm I have the same issue with `en_core_web_trf` on GPU. I have the following setting:

- I need to process a very long (millions of characters) text
- I split the text in chunks of, let's say 50k chars, so that each pipeline call doesn't take too much GPU memory. In this case it consumes around 1GB.
- I've used both the suggestion from @adrianeboyd to set the gpu allocator to pytorch, and also adding `with torch.no_grad()`. On top of that, I also call `gc.collect()` and `torch.cuda.empty_cache()` after each pipeline call.
- As a result, this does free up memory after processing each chunk, but not completely, so ~200MBs are accumulated after each pipeline call. Therefore, with a long enough text, I eventually get a CUDA OOM error.

**Comment by adrianeboyd** [comment]
Now that we set the pytorch allocator by default as of thinc v8.0.14, I think the original issue should be resolved.

**Comment by martin-kirilov** [comment]
@adrianeboyd adding the `doc_cleaner` solved the problem, thanks!
